test/domain/survey_test.py

```diff
@@ -0,0 +1,295 @@
+    def test_extends_quiz_attempt_with_any_new_answers(self, quiz_attempt_get_by_user_id_mock):
+        attempt = QuizAttempt(user_id=1, questions=[
+            QuizAttemptAnswer(question_number=1, answer=0),
+            QuizAttemptAnswer(question_number=2, answer=0),
+            QuizAttemptAnswer(question_number=3, answer=0),
+            QuizAttemptAnswer(question_number=4, answer=0),
+            QuizAttemptAnswer(question_number=5, answer=0),
+        ])
+        quiz_attempt_get_by_user_id_mock.return_value = attempt
+        self.survey.save_user_submitted_answers([
+            QuizAttemptAnswer(question_number=5, answer=5),
+            QuizAttemptAnswer(question_number=6, answer=0),
+            QuizAttemptAnswer(question_number=7, answer=5),
+            QuizAttemptAnswer(question_number=8, answer=2),
+            QuizAttemptAnswer(question_number=9, answer=2),
+            QuizAttemptAnswer(question_number=10, answer=5),
+        ])
+        self.assertEqual(5, attempt.questions[4].answer)
+        self.assertEqual(10, len(attempt.questions))
+
+    @mock.patch('app.domain.survey.QuizAttempt.get_by_user_id')
+    def test_answer_not_written_if_question_numbers_dont_match(self, quiz_attempt_get_by_user_id_mock):
+        attempt = QuizAttempt(user_id=1, questions=[
+            QuizAttemptAnswer(question_number=1, answer=0),
+            QuizAttemptAnswer(question_number=2, answer=0),
+            QuizAttemptAnswer(question_number=3, answer=0),
+        ])
+        quiz_attempt_get_by_user_id_mock.return_value = attempt
+        self.survey.save_user_submitted_answers([
+            QuizAttemptAnswer(question_number=1, answer=2),
+            QuizAttemptAnswer(question_number=3, answer=5),
+            QuizAttemptAnswer(question_number=4, answer=5),
+        ])
+        self.assertEqual(0, attempt.questions[1].answer)
+        self.assertEqual(5, attempt.questions[2].answer)
+        self.assertEqual(5, attempt.questions[3].answer)
+
+    def test_page_num_is_required(self):
+        with self.assertRaises(ValueError):
+            self.survey._get_questions_for_survey_page(None)
+
+    def test_page_num_must_be_positive_integer(self):
+        with self.assertRaises(ValueError):
+            self.survey._get_questions_for_survey_page(-1)
+
+    def test_page_num_can_not_be_string_one(self):
+        with self.assertRaises(ValueError):
+            self.survey._get_questions_for_survey_page("one")
+
+    @mock.patch('app.domain.survey.Survey._calculate_from_and_to_for_page_number')
+    def test_page_num_is_converted_to_integer(self, calculate_page_number_mock):
+        calculate_page_number_mock.return_value = (1, 15)
+        self.survey._get_questions_for_survey_page("1")
+        calculate_page_number_mock.assert_called_once_with(1)
+
+    @mock.patch('app.domain.survey.Question.get_questions_by_number_range')
+    def test_page_num_1_returns_questions_1_to_20(self, get_questions_by_number_range_mock):
```

Comment:
These might actually be worth deleting anyway. This functionality is tested in the calculate from and to for page number stuff anyway.